includes/class-plugin-autoupdate-filter-helpers.php

```diff
@@ -0,0 +1,138 @@
+		$update_version_parts    = explode( '.', $update_version );
+
+		if ( 3 !== count( $installed_version_parts ) || 3 !== count( $update_version_parts ) ) {
+			return false;
+		}
+
+		// only apply delays to major and minor releases. let point releases (patches) go through.
+		if ( $installed_version_parts[0] !== $update_version_parts[0] || $installed_version_parts[1] !== $update_version_parts[1] ) {
+			$update_allowed_after = $this->get_delay_date( $plugin_slug, $update_version, $delay_days );
+
+			if ( time() >= $update_allowed_after ) {
+				$this->clear_plugin_delay( $plugin_slug );
+				return true;
+			}
+
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieve the current version of an installed plugin.
+	 *
+	 * @param string $plugin_slug Slug of the plugin.
+	 *
+	 * @return string Current version of the plugin or an empty string if not found.
+	 */
+	public function get_installed_plugin_version( string $plugin_slug ): string {
+		if ( ! function_exists( 'get_plugins' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+
+		$plugins = get_plugins();
+		foreach ( $plugins as $plugin_path => $plugin_info ) {
+			if ( dirname( $plugin_path ) === $plugin_slug ) {
+				return $plugin_info['Version'];
+			}
+		}
+
+		return '';
+	}
+
+	/**
+	 * Retrieve the date after which a plugin update is allowed or calculate it if not set.
+	 *
+	 * @param string $plugin_slug Slug of the plugin.
+	 * @param string $update_version The version to update to.
+	 * @param int $delay_days Number of days to delay the update.
+	 *
+	 * @return int The Unix timestamp indicating when the plugin can be updated.
+	 */
+	public function get_delay_date( string $plugin_slug, string $update_version, int $delay_days ): int {
+		$option_key = 'plugin_update_delays';
+		$delays     = get_option( $option_key, array() );
+
+		if ( ! isset( $delays[ $plugin_slug ][ $update_version ] ) ) {
+			$release_date = $this->get_plugin_release_date( $plugin_slug );
+
+			if ( ! $release_date ) {
+				$release_date = time();
+			}
+
+			$delays[ $plugin_slug ][ $update_version ] = strtotime( "+{$delay_days} days", $release_date );
+			update_option( $option_key, $delays );
+		}
+
+		return $delays[ $plugin_slug ][ $update_version ];
+	}
+
+	/**
+	 * Retrieve the release date of a plugin based on its slug.
+	 *
+	 * @param string $plugin_slug Slug of the plugin.
+	 * @return int The Unix timestamp of the release date or the current time if not available.
+	 */
+	public function get_plugin_release_date( string $plugin_slug ): int {
+		$response = wp_safe_remote_get( "https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&slug={$plugin_slug}" );
+
+		if ( is_wp_error( $response ) ) {
+			return time();
+		}
+
+		$plugin_info = json_decode( wp_remote_retrieve_body( $response ), true );
+		if ( isset( $plugin_info['last_updated'] ) ) {
+			return strtotime( $plugin_info['last_updated'] );
```

Comment:
Just in case the returned is in a format that `strtotime` cannot translate:

```suggestion
			return strtotime( $plugin_info['last_updated'] ) ?? time();
```

Reply:
Not a bad idea to protect against things like that. Let's tease this out a little bit. 

When we hit this endpoint: https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&slug=woocommerce

The last updated date is returned like this: `"last_updated": "2024-04-30 11:42am GMT",` which is parsed with strtotime without a problem.

Brainstorming some possible ways that could maybe return something that strtotime couldn't handle:
- It's hosted on a 32-bit server, and the date is outside of the 32-bit unix epoch (i.e. any date after year 2038)
- There's an API change, and that date isn't included, or is renamed
- Anything else?

Reply:
Yes - if the plugin's date is written in a different format... I _think_ we've seen this happen in the woo marketplace before with paid plugins. while it's technically a bug, it could happen - especially with non-repo plugins.